flatten.cpp

```diff
@@ -0,0 +1,916 @@
+        printf("Rewriting history on top of llvm-project@%s...\n", buf);
+
+        // Disabled since the merged MLIR root commit has zero parents.
+        // Take the same approach to be consistent (= false).
+        const bool use_llvm_project_head_as_root = false;
+        if (use_llvm_project_head_as_root) {
+            new_head_renamed = &llvm_project_head;
+            is_root = false;
+        }
+
+        // Grab the llvm_project_tree.
+        git_commit *c;
+        check(git_commit_lookup(&c, repo, &llvm_project_head), "git_commit_lookup", NULL);
+        check(git_commit_tree(&llvm_project_tree, c), "git_commit_tree", NULL);
+        git_commit_free(c);
+
+        git_oid new_tree;
+
+        // For each commit, rewrite its tree.
+        while (!git_revwalk_next(&old_head, walk)) {
+            git_commit *c;
+            check(git_commit_lookup(&c, repo, &old_head), "git_commit_lookup", NULL);
+
+            insert_flang_directory(repo, &new_tree, git_commit_tree_id(c));
+
+            check(git_commit_create_from_ids(
+                &new_commit_id,
+                repo,
+                NULL,
+                git_commit_author(c),
+                git_commit_committer(c),
+                git_commit_message_encoding(c),
+                git_commit_message_raw(c),
+                &new_tree,
+                is_root ? 0 : 1,
+                (const git_oid**)(&new_head_renamed)
+            ), "git_commit_create_from_ids", NULL);
+            new_head_renamed = &new_commit_id;
+            is_root = false;
+
+            git_commit_free(c);
+        }
+
+        git_signature *merge_commit_author;
+        check(git_signature_default(&merge_commit_author, repo), "git_signature_default", NULL);
+
+        const char *merge_message =
+            "[Flang] Merge flang-compiler/f18\n"
+            "\n"
+            "This is the initial merge of flang-compiler, which is done in this way\n"
+            "principally to preserve the history and git-blame, without generating a large\n"
+            "number of commits on the first-parent history of LLVM.\n"
+            "\n"
+            "If you don't care about the flang history during a bisect remember that you can\n"
+            "supply paths to git-bisect, e.g. `git bisect start clang llvm`.\n"
+            "\n"
+            "The history of f18 was rewritten to:\n"
+            "\n"
+            "* Put the code under /flang/.\n"
+            "* Linearize the history.\n"
+            "* Rewrite commit messages so that issue and PR numbers point to the old repository.\n"
+            "\n"
+            "Updates: flang-compiler/f18#876 (submission into llvm-project)\n"
+            "Mailing-list: http://lists.llvm.org/pipermail/llvm-dev/2020-January/137989.html ([llvm-dev] Flang landing in the monorepo - next Monday!)\n"
+            "Mailing-list: http://lists.llvm.org/pipermail/llvm-dev/2019-December/137661.html ([llvm-dev] Flang landing in the monorepo)\n";
```

Comment:
@tskeith This is the commit message of the merge. I chose to include some extra references here so that people unfamiliar with what's going on have something to go by when they see the merge in their histories. Please review this, which is new.

Reply:
> @tskeith This is [the commit message](https://github.com/peterwaller-arm/f18/commit/527fdd0649249bd076681cf0c2230eb096832f7f) of the merge. I chose to include some extra references here so that people unfamiliar with what's going on have something to go by when they see the merge in their histories. Please review this, which is new.

That looks okay to me.